beacon-chain/state/state-native/hasher.go

```diff
@@ -0,0 +1,267 @@
+package state_native
+
+import (
+	"context"
+	"encoding/binary"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state/stateutil"
+	fieldparams "github.com/prysmaticlabs/prysm/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/config/params"
+	"github.com/prysmaticlabs/prysm/crypto/hash"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/encoding/ssz"
+	"github.com/prysmaticlabs/prysm/runtime/version"
+	"go.opencensus.io/trace"
+)
+
+// ComputeFieldRootsWithHasher hashes the provided state and returns its respective field roots.
+func ComputeFieldRootsWithHasher(ctx context.Context, state *BeaconState) ([][]byte, error) {
+	_, span := trace.StartSpan(ctx, "ComputeFieldRootsWithHasher")
+	defer span.End()
+
+	if state == nil {
+		return nil, errors.New("nil state")
+	}
+	hasher := hash.CustomSHA256Hasher()
+	var fieldRoots [][]byte
+	switch state.version {
+	case version.Phase0:
+		fieldRoots = make([][]byte, params.BeaconConfig().BeaconStateFieldCount)
+	case version.Altair:
+		fieldRoots = make([][]byte, params.BeaconConfig().BeaconStateAltairFieldCount)
+	case version.Bellatrix:
+		fieldRoots = make([][]byte, params.BeaconConfig().BeaconStateBellatrixFieldCount)
+	}
+
+	fieldRootIx := 0
+
+	// Genesis time root.
+	genesisRoot := ssz.Uint64Root(state.genesisTime)
+	fieldRoots[fieldRootIx] = genesisRoot[:]
+	fieldRootIx++
```

Comment:
instead of manually incrementing the index here, we can use field indexes directly here. Earlier I mentioned making these compile time vars, we could simply do the same here and use those vars.